Maestro/src/Microsoft.DotNet.Maestro.WebApi/Controllers/CommitPushedController.cs

```diff
@@ -82,5 +96,82 @@ public class CommitPushedController : ApiController
 
             Trace.TraceInformation("CommitPushed Complete");
         }
+
+        private void VerifyPayloadSignature(byte[] requestContent)
+        {
+            string payloadSignature = GetPayloadSignature();
+            string computedHash = ComputeHashHexdigest(requestContent);
+
+            if (!CryptographicEquals(payloadSignature, computedHash))
+            {
+                Trace.TraceWarning($"Payload signature: {payloadSignature}");
+                Trace.TraceWarning($"Computed hash: {computedHash}");
+
+                throw new Exception("The payload signature specified by the X-Hub-Signature header does not match " +
+                    "the hash computed from the request.  This could indicate that the repo's Maestro webhook has " +
+                    "an incorrect Secret, the request came from an unauthorized source, or the request was corrupted.");
+            }
+        }
+
+        private string GetPayloadSignature()
+        {
+            IEnumerable<string> headerValues;
+            string requestXHubSignature = string.Empty;
+            if (Request.Headers.TryGetValues("X-Hub-Signature", out headerValues))
+            {
+                requestXHubSignature = headerValues.FirstOrDefault();
+            }
+
+            if (string.IsNullOrWhiteSpace(requestXHubSignature) || !requestXHubSignature.StartsWith("sha1="))
+            {
+                throw new Exception("The HTTP request does not contain a valid X-Hub-Signature header.  Verify that " +
+                    "the repo's Maestro webhook is configured with a Secret.");
+            }
+
+            return requestXHubSignature.Substring(5).ToUpper();
+        }
+
+        private string ComputeHashHexdigest(byte[] requestContent)
+        {
+            byte[] hmacKey = Encoding.UTF8.GetBytes(Config.Instance.WebhookSecretToken);
```

Comment:
This is the same thing the asp.net webhooks repo (which I just learned about this morning) uses:

https://github.com/aspnet/WebHooks/blob/master/src/Microsoft.AspNet.WebHooks.Receivers.GitHub/WebHooks/GitHubWebHookReceiver.cs#L133